pixy/groupreg.go

```diff
@@ -0,0 +1,218 @@
+// `Topics()` channel. It also watches for other group members that join/leave
+// the group. Updated group member subscriptions are sent down to the
+// `MembershipChanges()` channel.
+type consumerGroupRegistry struct {
+	baseCID             *sarama.ContextID
+	group               string
+	config              *Config
+	groupZNode          *kazoo.Consumergroup
+	groupMemberZNode    *kazoo.ConsumergroupInstance
+	topicsCh            chan []string
+	membershipChangesCh chan map[string][]string
+	stoppingCh          chan none
+	wg                  sync.WaitGroup
+}
+
+func spawnConsumerGroupRegister(group, memberID string, config *Config, kazooConn *kazoo.Kazoo) *consumerGroupRegistry {
+	groupZNode := kazooConn.Consumergroup(group)
+	groupMemberZNode := groupZNode.Instance(memberID)
+	cgr := &consumerGroupRegistry{
+		baseCID:             sarama.RootCID.NewChild(fmt.Sprintf("registry:%s", group)),
+		group:               group,
+		config:              config,
+		groupZNode:          groupZNode,
+		groupMemberZNode:    groupMemberZNode,
+		topicsCh:            make(chan []string),
+		membershipChangesCh: make(chan map[string][]string),
+		stoppingCh:          make(chan none),
+	}
+	spawn(&cgr.wg, cgr.watcher)
+	spawn(&cgr.wg, cgr.register)
+	return cgr
+}
+
+func (cgr *consumerGroupRegistry) topics() chan<- []string {
+	return cgr.topicsCh
+}
+
+func (cgr *consumerGroupRegistry) membershipChanges() <-chan map[string][]string {
+	return cgr.membershipChangesCh
+}
+
+func (cgr *consumerGroupRegistry) claimPartition(cid *sarama.ContextID, topic string, partition int32, cancelCh <-chan none) func() {
+	if !retry(func() error { return cgr.groupMemberZNode.ClaimPartition(topic, partition) }, nil,
+		fmt.Sprintf("<%s> failed to claim partition", cid), cgr.config.Consumer.BackOffTimeout, cancelCh,
+	) {
+		log.Infof("<%s> partition claimed", cid)
+	}
+	return func() {
+		if !retry(func() error { return cgr.groupMemberZNode.ReleasePartition(topic, partition) },
+			func(err error) bool { return err != nil && err != kazoo.ErrPartitionNotClaimed },
+			fmt.Sprintf("<%s> failed to release partition", cid), cgr.config.Consumer.BackOffTimeout, cancelCh,
+		) {
+			log.Infof("<%s> partition released", cid)
+		}
+	}
+}
+
+func (cgr *consumerGroupRegistry) stop() {
+	close(cgr.stoppingCh)
+	cgr.wg.Wait()
+}
+
+// partitionOwner returns the id of the consumer group member that has claimed
+// the specified topic/partition.
+func (cgr *consumerGroupRegistry) partitionOwner(topic string, partition int32) (string, error) {
+	owner, err := cgr.groupZNode.PartitionOwner(topic, partition)
+	if err != nil {
+		return "", err
+	}
+	if owner == nil {
+		return "", nil
+	}
+	return owner.ID, nil
+}
+
+// watcher keeps an eye on the consumer group membership changes, that is when
+// members join/leave the group and sends notifications about such changes down
+// to the `membershipChangesCh` channel.
+func (cgr *consumerGroupRegistry) watcher() {
```

Comment:
This function is running in a gorotuine for entire `consumerGroupRegstry` life time. The very first change notification is to account for the membership change is to account for very first topic that the group member subscribes for. `register` goroutines subscribes to the topic and `watcher` detects that. So basically watcher is not only responsible for membership/subscription changes caused by other kafka-pixy instances but also by the instance it is a part of. I will add this to the function description to make it clear.